pydomcfg/domzgr/zgr.py

```diff
@@ -0,0 +1,160 @@
+"""
+Base class to generate NEMO v4.0 vertical grids.
+"""
+
+from itertools import product
+
+import xarray as xr
+from xarray import DataArray, Dataset
+
+
+class Zgr:
+    """
+    Base class to generate NEMO vertical grids.
+    """
+
+    def __init__(self, ds_bathy: Dataset, jpk: int):
+        """
+        Initialize class.
+
+        Parameters
+        ----------
+        ds_bathy: Dataset
+             xarray dataset including grid coordinates and bathymetry
+        jpk: int
+             number of computational levels
+        """
+
+        self._bathy = ds_bathy
+        self._jpk = jpk
+
+    # -------------------------------------------------------------------------
+    def _init_ds(self):
+        """
+        Initialise the xarray dataset with empty
+        ``z3{T,W}`` and ``e3{T,W}``
+
+        Returns
+        -------
+        ds: Dataset
+            A copy of the dataset used to initialise the class with new
+            coordinates ``z3{T,W}`` and ``e3{T,W}`` empty dataarrays
+
+        """
+        ds = self._bathy.copy()
+
+        var = ["z3", "e3"]
+        grd = ["T", "W"]
+
+        # Initialise a dataset with empty z3 and e3 dataarrays
+        da = xr.full_like(ds["Bathymetry"], None).expand_dims(z=range(self._jpk))
+        for v, g in product(var, grd):
+            ds[v + g] = da.copy()
+            ds = ds.set_coords(v + g)
+        return ds
+
+    # -------------------------------------------------------------------------
+    def _compute_sigma(self, kindx: DataArray):
+        """
+        Provide the analytical function for sigma-coordinate,
+        a uniform non-dimensional vertical coordinate describing
+        the non-dimensional position of model levels.
+
+        Consider that -1 <= sigma <= 0, with
+
+                 sigma =  0 at the shallower boundary
+                 sigma = -1 at the deeper boundary
+
+        Parameters
+        ----------
+        kindx: DataArray
+            Model levels indexes. Note that
+            *) T-points are at integer values (between 1 and jpk)
+            *) W-points are at integer values - 1/2 (between 0.5 and jpk-0.5)
+
+        Returns
+        -------
+        ps: tuple
+            (sigmaT, sigmaW) of Datarrays of uniform non-dimensional
+            sigma-coordinate (-1 <= sigma <= 0) for T and W grids
+        """
+
+        kindx = kindx + 1.0  # Fortran indexing
+
+        T = 0.5
+        W = 1.0
+        ps = (-(kindx - shift) / (self._jpk - 1.0) for shift in (T, W))
+
+        return ps
```

Comment:
Now that you translated most of it, isn't more natural to get rid of the line to align with Fortran indexing?
So W no shift, T +0.5 shift?
So just
```suggestion
        k_t = kindx + 0.5
        k_w = kindx
        ps = (-k / (self._jpk - 1.0) for kindx in (k_t, k_w))

        return ps
```